--- a/net/minecraft/client/gui/screen/EditBookScreen.java
+++ b/net/minecraft/client/gui/screen/EditBookScreen.java
@@ -5,6 +5,9 @@
 import com.mojang.blaze3d.systems.RenderSystem;
 import java.util.List;
 import java.util.ListIterator;
+
+import org.vivecraft.gameplay.screenhandlers.KeyboardHandler;
+
 import net.minecraft.client.gui.AbstractGui;
 import net.minecraft.client.gui.IGuiEventListener;
 import net.minecraft.client.gui.chat.NarratorChatListener;
@@ -124,6 +127,9 @@
             this.previousPage();
         }, true));
         this.updateButtons();
+        //Vivecraft
+        KeyboardHandler.setOverlayShowing(true);
+        //
     }
 
     private String removeUnprintableChars(String text)
