--- a/net/minecraft/item/CrossbowItem.java
+++ b/net/minecraft/item/CrossbowItem.java
@@ -5,6 +5,11 @@
 import java.util.Random;
 import java.util.function.Predicate;
 import javax.annotation.Nullable;
+
+import org.vivecraft.api.NetworkHelper;
+import org.vivecraft.api.ServerVivePlayer;
+import org.vivecraft.utils.math.Vector3;
+
 import net.minecraft.advancements.CriteriaTriggers;
 import net.minecraft.client.renderer.Quaternion;
 import net.minecraft.client.renderer.Vector3f;
@@ -279,12 +284,20 @@
             }
             else
             {
-                Vec3d vec3d1 = shooter.getUpVector(1.0F);
-                Quaternion quaternion = new Quaternion(new Vector3f(vec3d1), projectileAngle, true);
-                Vec3d vec3d = shooter.getLook(1.0F);
-                Vector3f vector3f = new Vector3f(vec3d);
-                vector3f.transform(quaternion);
-                iprojectile.shoot((double)vector3f.getX(), (double)vector3f.getY(), (double)vector3f.getZ(), velocity, inaccuracy);
+            	Vec3d vec3d1 = shooter.getUpVector(1.0F);
+            	Vec3d vec3d = shooter.getLook(1.0F);
+            	//VIVE SET AIM AND SPEED
+            	ServerVivePlayer vp = NetworkHelper.vivePlayers.get(shooter.getUniqueID());
+            	if (vp !=null && vp.isVR())
+            	{ 
+            		vec3d = vp.getControllerDir(vp.activeHand);	
+            		vec3d1 = vp.getControllerVectorCustom(vp.activeHand, new Vector3(0, 1, 0));
+            	}
+            	///              
+            	Vector3f vector3f = new Vector3f(vec3d);
+            	Quaternion quaternion = new Quaternion(new Vector3f(vec3d1), projectileAngle, true);
+            	vector3f.transform(quaternion);
+            	iprojectile.shoot((double)vector3f.getX(), (double)vector3f.getY(), (double)vector3f.getZ(), velocity, inaccuracy);
             }
 
             crossbow.damageItem(flag ? 3 : 1, shooter, (p_220017_1_) ->
